### PR TITLE
UIViewController lifecycle signals

### PR DESCRIPTION
**Pull request by jdmoreira**
Hello friends 👋
I've noticed you were discussing adding signals for the `UIViewController` lifecycles methods.
Since I already had an implementation on my hard drive, I've decided it was better to just send a pull request.

### What
Add signals to `UIViewController` so we can listen on lifecycle events.

### How
Swizzling the lifecycle methods and calling a `callbacker` that lives inside `UIViewController` as an associated object.

### CODE REVIEW

**Comment by mansbernhardt** [comment]
@jdmoreira I like this solution using an extension on UIViewController instead of a sub-class, as it's much more usable and much less intrusive to the users design as it can be used on any of an user's subclasses without requiring using our sub-class as suggested in https://github.com/iZettle/Flow/pull/40.

I must admit that the swizzling makes me a bit worried, but I guess it's a proven solution to similar problems? I know that you did investigate doing something similar for `UIView`'s helpers such as `UIView.didLayoutSignal` but you ran into problems in that case? Do you remember why it could not be applied to UIView as well? I think that would be nice than our current solution.

**Comment by mansbernhardt** [comment]
@jdmoreira Any reason for signaling the current view in all the signals above? The current view should be should be easy accessible from the context when needed.

```swift
bag += vc.viewDidLoadSignal.onValue {
  vc.view 
}
```

I think in general that a signal on an instance should not signal the instance itself or anything easily obtainable from that instance.

**Comment by mansbernhardt** [comment]
@jdmoreira In https://github.com/iZettle/Flow/pull/40#issuecomment-426937856, I showed another issue when working directly with `UIViewController` where subclassing is required to customize its behavior such as `preferredStatusBarStyle` etc. Do you think swizzling similar as above would be a way forward to solve that as well instead of falling back on a subclass?

**Comment by jdmoreira** [comment]
Hello Måns!

I think when we last discussed it, it was as an alternative to the `CallbackerView` that we used for the `hasWindowSignal`. At that time we weren't sure what was more performant, adding a subview or swizzling and we ended up not using the swizzling implementation.

For sure swizzling does feel hacky but I tend to use it a lot in my own projects since it's such a convenient tool. I think it would allow you to remove the subclass discussed in iZettle/Flow#40 as well.

There's a lot of swift code using swizzling out there (`RxSwift` for example) and there will probably be code using it for as long as we have `NSObject`. I believe swizzling itself is quite performant since from my understanding it just changes the entry in the dispatch directly.

In this implementation besides the swizzling overhead (I believe it's zero) there is the overhead of calling the `callbacker` and filtering the signal per `UIViewController` lifecycle method - probably not that expensive either 🤔 

Regarding the type of the signals being `Signal<UIView>` I agree with you, there is no point in returning the view. We can always default to `Void`. The ideal would be to return the `animated` parameter for some of the signals... It would make the code slightly more complex but it's probably worth it.

**Comment by mansbernhardt** [comment]
I wonder if we can approach the implementation a bit like like we do we our current delegates. The idea is to keep a private delegate instance as an associated object, then swizzle the methods to call the delegates callback or delegates:

```swift
private class Delegate {
   let viewDidLoad = Callbacker<()>()
   let viewWillAppear = Callbacker<Bool>()
   var targetViewController = Delegate<(action: Selector, sender: Any?), UIViewController?>()
}
```

```swift
extension UIViewController {
  let viewDidLoadSignal: Signal<()> { return Signal(backer: delegate. viewDidLoad) }
  let targetViewController:  Delegate<(action: Selector, sender: Any?), UIViewController?> { return delegate. targetViewController }
}
```

```swift
extension UIViewController {
    @objc func _swizzled_viewDidLoad() {
         self._swizzled_viewDidLoad()
         delegate.viewDidLoad.callAll()
     }

    @objc func _swizzled_viewWillAppear(_ animated: Bool) {
         self._swizzled_viewDidAppear(animated)
         delegate.viewWillAppear.callAll(with: animated)
     }

    @objc func _swizzled_ targetViewController(forAction action: Selector, sender: Any?) -> UIViewController? {
         return delegate.targetViewController.call((action, sender)) ?? nil 
    }
}
```

**Comment by mansbernhardt** [comment]
I guess, as for the styling/configuration parts such as `preferredStatusBarStyle` could be implemented separably using another associated object for the current style, and we can setup another bunch of swizzling for these?

**Comment by jdmoreira** [comment]
Yes, we could do it that way and keep a call backer for each method. I'll give it a try.

Regarding the swizzling for `preferredStatusBarStyle` there is two different approaches that I can think of. One is to do all the swizzling for a class at once, something like this here: https://gist.github.com/jdmoreira/04bf23be26bbd0bd9f042b36b275e14b

The other way would be to swizzle once at a time - different frameworks for example - and then we should probably not rely on a protocol like `Swizzable`.

File: Presentation/UIViewController+Signals.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+public extension UIViewController {
+    var viewDidLoadSignal: Signal<Void> {
+        return Signal(callbacker: callbackers.viewDidLoad)
+    }
+
+    var viewWillAppearSignal: Signal<Bool> {
+        return Signal(callbacker: callbackers.viewWillAppear)
+    }
+
```
Comment (CarlEkman) [inline Presentation/UIViewController+Signals.swift]: Looks like `viewDidAppear` is missing :) There is of course also the callback `didReceiveMemoryWarning`, but maybe it doesn't belong here.

Comment (jdmoreira) [reply Presentation/UIViewController+Signals.swift]: Good catch. I will add `viewDidAppear`. Not sure about `didReceiveMemoryWarning`, maybe worth it?

Comment (mansbernhardt) [reply Presentation/UIViewController+Signals.swift]: Low memory can be handled via notifications instead:

```swift
bag += NotificationCenter.default.signal(forName: UIApplication.didReceiveMemoryWarningNotification).onValue { ... }
```

**Comment by jdmoreira** [comment]
Thanks for the reviews, closing this one now. Continuing it in Flow at 
https://github.com/iZettle/Flow/pull/56

**Pull request closed without merging**